File: web_version/pmugg_files/guidelines/lineSegment.js
ms.lineSegment = function(stats) {
	var properties = {
		lineState: new ms.requiredArray('lineState'),
		line: new ms.singleProperty('line', /* required */ true),
	};
	this.node = new ms.node(this, stats, 'lineSegment', properties);

	// The distance lower bound is the distance when it's below the 
	// ideal distance, beyond that the distance is not calculated.
	this.lengthLowerBound = 0;
	
	// Dirty length is currently not used.
	this.dirtyLength = true;
};

ms.lineSegment.moveMutable = false;

ms.lineSegment.prototype.getNode = function() {
	return this.node;
};

ms.lineSegment.prototype.getStates = function() {
	return this.node.get('lineState');
};

ms.lineSegment.prototype.getState = function(isAtStart) {
	var states = this.getStates();
	return states[isAtStart ? 0 : states.length - 1];
};

ms.lineSegment.prototype.getLine = function() {
	return this.node.get('line');
};

ms.lineSegment.prototype.addState = function(state, insertAtStart) {
	this.dirtyLength = true;
	this.node.connect(state, insertAtStart);
};

ms.lineSegment.prototype.addStates = function(states) {
	this.dirtyLength = true;
	var self = this;
	states.forEach(function(state) {
		self.node.connect(state);
	});
};

ms.lineSegment.prototype.destroy = function() {
	this.getStates().forEach(function(state) {
		state.getNode().destroy();
	});
};

ms.lineSegment.prototype.merge = function(segmentB, mergedState) {
	var statesA = this.getStates().slice();
	var statesB = segmentB.getStates().slice();
	// Throw out the two states that were merged and put in the mergedState.
	statesA[statesA.length - 1].getNode().disconnect(this);
	statesB.shift();
	statesB.unshift(mergedState);
	this.addStates(statesB);
	segmentB.node.disconnect(segmentB.getLine());
	segmentB.node.destroy();
};

ms.lineSegment.prototype.split = function(unsplitState, splitPoint, vertex, opt_index) {
	ms.timerG.start('split A');
	this.dirtyLength = 0;
	var index = opt_index || this.getStates().indexOf(unsplitState);
	if (index == -1) {
		ms.alert('Cannot find state to split.');
	}

	var newSegment = new ms.lineSegment(this.node.getStats());
	newSegment.addStates(this.getStates().slice(index + 1, this.getStates().length));

	var splitStates = unsplitState.split(splitPoint);
	this.addState(splitStates[0], false);
	newSegment.addState(splitStates[1], true);
	unsplitState.destroy();

	var line = this.getLine();
	var segmentIndex = line.getSegments().indexOf(this);
	line.getNode().splice(newSegment, segmentIndex + 1);

	vertex.addLineState(splitStates[0], 0);
	vertex.addLineState(splitStates[1], 1);
	ms.timerG.stop('split A');
};

// TODO: Combine with setPosition.
// This remove the line segment from the cells, but does not add them back.
ms.lineSegment.prototype.setPositionsOneState = function(positions) {
	var states = this.getStates();
	while (states.length > 1) {
		var state = states[states.length - 1];
		state.getNode().disconnect(this);
		state.destroy();
	}	
	states[0].setPosition(positions[0], true);
	states[0].setPosition(positions[1], false);
	states[0].refreshCells();
};

// TODO: Combine with setPosition.
// This remove the line segment from the cells, but does not add them back.
ms.lineSegment.prototype.setPositions = function(positions) {
	var states = this.getStates().filter(function(state) {
		return !ms.lineSegment.moveMutable || state.isMutable();		
	});	
	
	var p0 = states[0].getPosition(true);
	var p1 = states[states.length - 1].getPosition(false);
	var q0 = positions[0];
	var q1 = positions[1];
	var ua = p1.copy().minus(p0);
	var ub = q1.copy().minus(q0);
	var va = new ms.vec2(-ua.y, ua.x);
	var vb = new ms.vec2(-ub.y, ub.x);
	var A = new ms.matrix('full', [[ua.x, va.x], [ua.y, va.y]]);
	var B = new ms.matrix('full', [[ub.x, vb.x], [ub.y, vb.y]]);
	var P0 = new ms.matrix('full', [[p0.x], [p0.y]]);
	var BAi = B.dot(A.inv());
	var BAip0 = BAi.dot(P0);
	var offset = new ms.vec2(-BAip0.x() + q0.x, -BAip0.y() + q0.y);
	
	for (var i = 0; i < states.length; i++) {
		for (var atEnd = 0; atEnd < 2; atEnd++) {
			var p = states[i].getPosition(atEnd == 0);
			var P = new ms.matrix('full', [[p.x], [p.y]]);
			var BAip = BAi.dot(P);
			var newPosition = new ms.vec2(BAip.x() + offset.x, BAip.y() + offset.y);
			states[i].setPosition(newPosition, atEnd == 0);
			states[i].refreshCells();
		}
	}
};

ms.lineSegment.prototype.setPosition = function(newPosition, isAtStart) {
	var totalLength = 0;
	var cumulativeLengths = [0];
	var states = this.getStates().filter(function(state) {
		return !ms.lineSegment.moveMutable || state.isMutable();		
	});
	states.forEach(function(state) {
		totalLength += state.getLength();
		cumulativeLengths.push(totalLength);
	});
	if (totalLength == 0) {
		ms.alert('Zero mutable length.');
	}
	
	var state = states[isAtStart ? 0 : states.length - 1];
	var position0 = state.getPosition(isAtStart);
	var fullMovement = newPosition.copy().minus(position0);
	var movements = [];
	for (var i = 0; i < cumulativeLengths.length; i++) {
		var scale = (isAtStart ? totalLength - cumulativeLengths[i] : cumulativeLengths[i]) / totalLength;
		movements.push(fullMovement.copy().scale(scale));
	}
	for (var i = 0; i < states.length; i++) {
		states[i].move(movements[i], movements[i + 1]);
	}
};

ms.lineSegment.prototype.findIntersections = function() {
	return this.getStates().some(function(state) {
		return state.countIntersections() > 0;
	});
};

ms.lineSegment.prototype.getLength = function(idealLength) {
	ms.timerG.start('Segment Length');
	// Dirty length could be used here to speed things up, but some states may become immutable.
	// if (this.dirtyLength) {
	this.lengthLowerBound = 0;
	var states = this.getStates();
	for (var i = 0; i < states.length; i++) {
		if (!ms.lineSegment.moveMutable || states[i].isMutable()) {
			this.lengthLowerBound += states[i].getLength();
			if (this.lengthLowerBound >= idealLength) {
				break;
			}
		}
	}
	// 	this.dirtyLength = false;
	// }
	ms.timerG.stop('Segment Length');
	return this.lengthLowerBound;
};

ms.lineSegment.prototype.highlight = function(context, convertToScreen) {
	for (var i = 0; i < this.getStates().length; i++) {
		this.getStates()[i].highlight(context, convertToScreen);
	}
};

ms.lineSegment.prototype.print = function() {
	var states = this.getStates();
	window.console.log(states.length + ' states');
	states[0].print();
	states[states.length - 1].print();
	ms.highlight(this);
};

File: cpp_version/pmugg/guidelines/line_segment.cpp
#include "line_segment.h"
#include "line_state.h"
#include "line.h"
#include "vertex.h"
#include "timer.h"
#include "util.h"
#include <algorithm>

namespace ms {

bool LineSegment::moveMutable = false;

LineSegment::LineSegment(Stats* stats)
    : lengthLowerBound(0)
    , dirtyLength(true) {
    
    std::map<std::string, std::unique_ptr<Property>> properties;
    properties["lineState"] = std::make_unique<RequiredArray>("lineState");
    properties["line"] = std::make_unique<SingleProperty>("line", true);

    node = std::make_unique<Node>(this, stats, "lineSegment", std::move(properties));
}

Node* LineSegment::getNode() const {
    return node.get();
}

std::vector<LineState*> LineSegment::getStates() const {
    return node->get("lineState");
}

LineState* LineSegment::getState(bool isAtStart) const {
    auto states = getStates();
    return states[isAtStart ? 0 : states.size() - 1];
}

Line* LineSegment::getLine() const {
    return node->get("line");
}

float LineSegment::getLength(float idealLength) {
    timer->start("Segment Length");
    
    // Dirty length could be used here to speed things up, but some states may become immutable
    lengthLowerBound = 0;
    auto states = getStates();
    
    for (auto* state : states) {
        if (!moveMutable || state->isMutable()) {
            lengthLowerBound += state->getLength();
            if (lengthLowerBound >= idealLength) {
                break;
            }
        }
    }
    
    timer->stop("Segment Length");
    return lengthLowerBound;
}

bool LineSegment::findIntersections() {
    return std::any_of(getStates().begin(), getStates().end(),
        [](LineState* state) { return state->countIntersections() > 0; });
}

void LineSegment::addState(LineState* state, bool insertAtStart) {
    dirtyLength = true;
    node->connect(state, insertAtStart);
}

void LineSegment::addStates(const std::vector<LineState*>& states) {
    dirtyLength = true;
    for (auto* state : states) {
        node->connect(state);
    }
}

void LineSegment::destroy() {
    for (auto* state : getStates()) {
        state->getNode()->destroy();
    }
}

void LineSegment::merge(LineSegment* segmentB, LineState* mergedState) {
    auto statesA = getStates();
    auto statesB = segmentB->getStates();
    
    // Throw out the two states that were merged and put in the mergedState
    statesA.back()->getNode()->disconnect(this);
    statesB.front()->getNode()->disconnect(segmentB);
    statesB[0] = mergedState;
    
    addStates(statesB);
    segmentB->node->disconnect(segmentB->getLine());
    segmentB->node->destroy();
}

void LineSegment::split(LineState* unsplitState, const Vec2& splitPoint, 
                       Vertex* vertex, int index) {
    timer->start("split A");
    dirtyLength = true;
    
    if (index == -1) {
        index = std::find(getStates().begin(), getStates().end(), unsplitState) - 
                getStates().begin();
    }
    
    if (index == -1) {
        ms::alert("Cannot find state to split.");
    }

    auto* newSegment = new LineSegment(node->getStats());
    newSegment->addStates(std::vector<LineState*>(
        getStates().begin() + index + 1,
        getStates().end()
    ));

    auto splitStates = unsplitState->split(splitPoint);
    addState(splitStates[0], false);
    newSegment->addState(splitStates[1], true);
    unsplitState->destroy();

    auto* line = getLine();
    auto segmentIndex = std::find(line->getSegments().begin(), 
                                 line->getSegments().end(), 
                                 this) - line->getSegments().begin();
    line->getNode()->splice(newSegment, segmentIndex + 1);

    vertex->addLineState(splitStates[0], 0);
    vertex->addLineState(splitStates[1], 1);
    
    timer->stop("split A");
}

void LineSegment::setPositionsOneState(const std::vector<Vec2>& positions) {
    auto states = getStates();
    while (states.size() > 1) {
        auto* state = states.back();
        state->getNode()->disconnect(this);
        state->destroy();
        states.pop_back();
    }
    
    states[0]->setPosition(positions[0], true);
    states[0]->setPosition(positions[1], false);
    states[0]->refreshCells();
}

void LineSegment::setPositions(const std::vector<Vec2>& positions) {
    auto states = getStates();
    states.erase(
        std::remove_if(states.begin(), states.end(),
            [](LineState* state) { return !moveMutable || state->isMutable(); }),
        states.end()
    );
    
    if (states.empty()) return;
    
    auto p0 = states.front()->getPosition(true);
    auto p1 = states.back()->getPosition(false);
    auto q0 = positions[0];
    auto q1 = positions[1];
    
    auto ua = p1 - p0;
    auto ub = q1 - q0;
    auto va = Vec2(-ua.y, ua.x);
    auto vb = Vec2(-ub.y, ub.x);
    
    Matrix A({ua.x, va.x, ua.y, va.y}, 2, 2);
    Matrix B({ub.x, vb.x, ub.y, vb.y}, 2, 2);
    Matrix P0({p0.x, p0.y}, 2, 1);
    
    auto BAi = B * A.inverse();
    auto BAip0 = BAi * P0;
    auto offset = Vec2(-BAip0.get(0, 0) + q0.x, -BAip0.get(1, 0) + q0.y);
    
    for (auto* state : states) {
        for (int atEnd = 0; atEnd < 2; atEnd++) {
            auto p = state->getPosition(atEnd == 0);
            Matrix P({p.x, p.y}, 2, 1);
            auto BAip = BAi * P;
            auto newPosition = Vec2(BAip.get(0, 0) + offset.x, BAip.get(1, 0) + offset.y);
            state->setPosition(newPosition, atEnd == 0);
            state->refreshCells();
        }
    }
}

void LineSegment::setPosition(const Vec2& newPosition, bool isAtStart) {
    auto states = getStates();
    states.erase(
        std::remove_if(states.begin(), states.end(),
            [](LineState* state) { return !moveMutable || state->isMutable(); }),
        states.end()
    );
    
    if (states.empty()) return;
    
    float totalLength = 0;
    std::vector<float> cumulativeLengths = {0};
    
    for (auto* state : states) {
        totalLength += state->getLength();
        cumulativeLengths.push_back(totalLength);
    }
    
    if (totalLength == 0) {
        ms::alert("Zero mutable length.");
        return;
    }
    
    auto* state = isAtStart ? states.front() : states.back();
    auto position0 = state->getPosition(isAtStart);
    auto fullMovement = newPosition - position0;
    
    std::vector<Vec2> movements;
    for (float length : cumulativeLengths) {
        float scale = (isAtStart ? totalLength - length : length) / totalLength;
        movements.push_back(fullMovement * scale);
    }
    
    for (size_t i = 0; i < states.size(); i++) {
        states[i]->move(movements[i], movements[i + 1]);
    }
}

void LineSegment::highlight(void* context, const std::function<Vec2(const Vec2&)>& convertToScreen) {
    for (auto* state : getStates()) {
        state->highlight(context, convertToScreen);
    }
}

void LineSegment::print() const {
    auto states = getStates();
    std::cout << states.size() << " states" << std::endl;
    states.front()->print();
    states.back()->print();
    ms::highlight(this);
}

void LineSegment::updateLength() {
    if (!dirtyLength) return;
    
    lengthLowerBound = 0;
    for (auto* state : getStates()) {
        if (!moveMutable || state->isMutable()) {
            lengthLowerBound += state->getLength();
        }
    }
    
    dirtyLength = false;
}

} // namespace ms 